src/runtime_src/core/pcie/emulation/common_em/config.h

```diff
@@ -89,6 +90,89 @@ namespace xclemulation{
   const int VIVADO_MIN_VERSION = 2000;
   const int VIVADO_MAX_VERSION = 2100;
 
+//sparse log utility
+enum class eEmulationType{
+  eSw_Emu,
+  eHw_Emu
+};
+
+struct sParseLog
+{
+  std::ifstream mFileStream;
+  std::string mFileName;
+  std::atomic_bool mFileExists;
+  std::vector<std::string> mMatchedStrings;
+  eEmulationType mEmuType;
+
+  sParseLog(const std::string& iDeviceLog, eEmulationType iType, const std::vector<std::string>& iMatchedStrings)
+      : mFileName(iDeviceLog)
+      , mFileExists{false}
+      , mMatchedStrings(iMatchedStrings)
+      , mEmuType(iType)
+  {
+
+  }
+
+  void check_simulator_status()
+  {
+    std::string line;
+    while (std::getline(mFileStream, line))
+    {
+      for (const auto &StringData : mMatchedStrings)
+      {
+        if (line.find(StringData) != std::string::npos)
+        {
+          if (eEmulationType::eSw_Emu == mEmuType)
+            print_user_msg();
+          else if (eEmulationType::eHw_Emu == mEmuType)
+          {
+            if (!StringData.compare("Exiting xsim") || !StringData.compare("FATAL_ERROR"))
+              print_user_msg();
+            else
+              std::cout << line << '\n';
+          }
+        }
+      }
+    }
+  }
+
+  void print_user_msg()
+  {
+    switch (mEmuType) {
+      case eEmulationType::eSw_Emu:
+        std::cout << "Received request to end the application. Press Cntrl+C to exit the application." << std::endl;
+        break;
+      case eEmulationType::eHw_Emu:
+        std::cout << "SIMULATION EXITED" << std::endl;
+        break;
+    }
+  }
+  /*********************************************************************************
+   *  The function traverse the log file (simulate.log) and prints an actionable 
+   *  user message if anyline of log file matches exactly with an user defined
+   *  vector of strings. The log file might be updated by a separate process. So ensuring
+   *  file existence before opening it would eliminate several exceptions.
+   *  
+   * */
+  void parseLog()
+  {
+    if (not mFileExists.load())
+    {
+      if (boost::filesystem::exists(mFileName))
+      {
+        mFileStream.open(mFileName);
+        if (mFileStream.is_open())
+          mFileExists.store(true);
+      }
+    }
+
+    if (mFileExists.load())
+    {
+      check_simulator_status();
+    }
+  }
+};
```

Comment:
Thank you, much clearer now.  It's amazing what proper function names can communicate.  The only recommendation I have  is to move the comment to appropriate places within the code.  For example, add the comment "the log file might be updated ..., so ensuring ..." to above line 159.   and the comment about traversal and user strings to within check_simulator_status().